First pass at tests for DiceRollParser.

toonarmycaptain opened the pull request.

toonarmycaptain committed:
First pass at tests for DiceRollParser.

toonarmycaptain commented on test/test_example_parsers/test_dice_roll_parser.py:
```diff
@@ -0,0 +1,65 @@
+from plusminus.examples.example_parsers import DiceRollParser
+
+
+@pytest.fixture
+def die():
+    return DiceRollParser()
```
This could be in a conftest.py, but since you said you might spin one or more of these off into their own repo, I've kept it with the tests for the parser. Pytest will search in the test script first, then elsewhere.

ptmcg replied:
I think you'd keep conftest.py reserved for those test features that are not parser-specific. Not sure I understand the value of this fixture, but it is probably a pytest idiom.

toonarmycaptain replied:
In this case, it's not doing much work except saving a couple of lines of setup, and if there's an error building the parser, then the error is highlighted as an error (in the fixture) rather than a test fail.
eg if you try:
```python3
@pytest.fixture
def die():
    import random
    if random.randint(0, 10)  % 2 == 0:
        raise ValueError
    return DiceRollParser()
```
and then run `python -m pytest -vv` for verbose output, you see all the PASS/FAIL/ERROR codes, except you won't see any failing tests. 

Writing it this way also lets you change the call if the name or something about `DiceRollParser` changes, say you need to pass an argument to it in the 'new version', you can just change it in the fixture rather than everywhere in the code. 

I believe you could scope it to be class or module scoped, such that it would cache the parser object rather than passing a new object for each case, which is what I believe it is doing now, which will get you some speedup, or would be useful if you wanted to use a fresh database for a series of tests.

toonarmycaptain commented on test/test_example_parsers/test_dice_roll_parser.py:
```diff
@@ -0,0 +1,65 @@
+                                    num_rolls):
+        assert num_rolls <= die.evaluate(f'{num_rolls}d1') == num_rolls
+
+    @pytest.mark.parametrize('num_rolls',
+                             [1, 2, 4, 5, 10, 100, ],
+                             )
+    @pytest.mark.parametrize('die_max',
+                             [2, 4, 5, 10, 100, ]
+                             )
```
You can usefully stack parametrizations, so can pass (1,2) and (3,4) and get (1,3), (1,4), (2,3), (2,4).
I just found this out tonight, but I like it very much!

toonarmycaptain commented on test/test_example_parsers/test_dice_roll_parser.py:
```diff
@@ -0,0 +1,65 @@
+                             )
+    def test_single_roll_ranges(self, die,
+                                num_rolls, die_max):
+        assert num_rolls <= die.evaluate(f'{num_rolls}d{die_max}') <= num_rolls*die_max
```
I don't know how useful this test and the following class are for this parser, unless one repeats the tests enough times to at least have a good chance of producing an edge or out of range result. The tests do confirm that the parser evaluates the input successfully, and that it is within the range of expected outputs, which is marginally more than the [original set of tests](https://github.com/pyparsing/plusminus/blob/1cd60b8f802e9b77c166cf5339437612b6647bde/plusminus/examples/example_parsers.py#L125-L127).

toonarmycaptain commented on test/test_example_parsers/test_dice_roll_parser.py:
```diff
@@ -0,0 +1,65 @@
+    def test_compound_examples(self, die,
+                               evaluation_string,
+                               result_min, result_max):
+        assert result_min <= die.evaluate(evaluation_string) <= result_max
```
Not sure how useful these tests end up being, aside from essentially testing the underlying ArithmeticParser, which ought to be covered by it's own tests, again with the utility being limited more with more complex examples with longer possible ranges.

ptmcg commented:
Make a test that rolls a 6-sided die 100 times, and checks that no result is outside the range 1-10 inclusive, and that each value 1-6 occurs at least once. Probably about the best you can do. I wouldn't bother checking the frequency of the values, at that point you are just testing the randomness of Python's builtin random() method. (I don't think this test will fit within the parametrize decorator though.)

The other thing to test is the working of 'd' as a binary operator. For instance, if rolling "2d6", you need to be sure this is the sum of 2 random rolls, not 2 times the value of a single roll. Do 100 (or however many) rolls of "2d6", and make sure all the values in the range 2-12 are present, and no values outside that range.

toonarmycaptain commented:
Alright. I looked at some more complicated tests, but knew some of them would be long running or non-deterministic. 100 rolls of `2d6` is not guaranteed to throw out every value in 2-12, for example. I don't know how you feel about that. There's probably some mark you can put on the non-deterministic ones to retry them if they fail, or some config you can pass to retry failing tests more generally. 
I actually experimented with how many tests until one of the compound examples actually hit the edge of the range, and ran a thousand or more without hitting the edge.

ptmcg commented:
Probably sufficient then to do 2d6 rolls and see if you come up with any odd number - at least then you would be sure you were doing the rolls correctly (unlike my first round which *did* do 2 X some random value, so all the rolls were even!). At that point, just assume that Python randint is doing the right thing.

toonarmycaptain committed:
Add test ensuring multiple rolls are distinct.

Test that multiple rolls are distinct rolls of the die, rather than a single
roll of the die multiplied by the number of rolls.

toonarmycaptain commented:
Well I guess I'll walk back my not guaranteed statement slightly - I tried 1000 runs of running `2d6` 10 times, then seeing if any of the results of the 10 was an odd number, and none of that 1000 failed, so I guess running 10x is a good bet. I've put it at 100 for now, taking ~0.75s, whereas 10 took ~0.78s, if that becomes a concern.

ptmcg commented on test/test_example_parsers/test_dice_roll_parser.py:
```diff
@@ -63,3 +63,9 @@ def test_compound_examples(self, die,
+    def test_rolls_are_distinct(self, die):
+        pairs_of_rolls = []
+        for pair in range(100):
+            pairs_of_rolls.append(die.evaluate('2d6'))
```
List comprehension?
22 should be 2
Add error message for asserrts

     pairs_of_rolls = [die.evaluate('2d6') for _ in range(100)]
     assert any(pair % 2 != 0 for pair in pairs_of_rolls), "incorrect evaluation of multiple die rolls"

toonarmycaptain replied:
Gah. I retyped after shortening the var.

toonarmycaptain replied:
You're right, a list comp is better. Actually, you could lose the comparison, if you wanted, and just have `pair%2 for pair in pairs of rolls` - but I don't think that's as clear to read.

ptmcg reviewed (commented):
Please add messages on asserts:

    assert condition_that_should_be_true, "description of failure in condition"

No f-strings, please. Target min release is Py3.5.

ptmcg commented:
Oh I was thinking of pyparsing, I never did specify what our minimum target Python release would be for this package.  Even so, I think 3.5 is a good target version to aim for. It is a balance between trying to be inclusive vs. using important language features. f-strings are just not that important a feature for me, so I don't feel that compelled to require people to be at 3.6 or later to use plusminus.

toonarmycaptain committed:
Fix assertion to test even, not divisible by 22, convert to list comprehension.

toonarmycaptain committed:
Fix casing typo.

toonarmycaptain committed:
Remove superfluous inequality in assertion.

toonarmycaptain committed:
Asdd error messages for asserts.

toonarmycaptain commented:
No problem. I don't think it's really worthwhile inserting the test case values in failure messages is really worth it for these tests anyway, which is the only reason I can see using f-strings here.
